src/server/pkg/serviceenv/service_env.go

```diff
@@ -0,0 +1,198 @@
+}
+
+// InitPachOnlyEnv initializes this service environment. This dials a GRPC
+// connection to pachd only (in a background goroutine), and creates the
+// template pachClient used by future calls to GetPachClient.
+//
+// This call returns immediately, but GetPachClient will block
+// until the client is ready.
+func InitPachOnlyEnv(config *Configuration) *ServiceEnv {
+	env := &ServiceEnv{Configuration: config}
+	env.pachAddress = net.JoinHostPort("localhost", fmt.Sprintf("%d", env.PeerPort))
+	env.pachEg.Go(env.initPachClient)
+	return env // env is not ready yet
+}
+
+// InitServiceEnv initializes this service environment. This dials a GRPC
+// connection to pachd and etcd (in a background goroutine), and creates the
+// template pachClient used by future calls to GetPachClient.
+//
+// This call returns immediately, but GetPachClient and GetEtcdClient block
+// until their respective clients are ready.
+func InitServiceEnv(config *Configuration) *ServiceEnv {
+	env := InitPachOnlyEnv(config)
+	env.etcdAddress = fmt.Sprintf("http://%s", net.JoinHostPort(env.EtcdHost, env.EtcdPort))
+	env.etcdEg.Go(env.initEtcdClient)
+	return env // env is not ready yet
+}
+
+// InitWithKube is like InitServiceEnv, but also assumes that it's run inside
+// a kubernetes cluster and tries to connect to the kubernetes API server.
+func InitWithKube(config *Configuration) *ServiceEnv {
+	env := InitServiceEnv(config)
+	env.kubeEg.Go(env.initKubeClient)
+	return env // env is not ready yet
+}
+
+func (env *ServiceEnv) initPachClient() error {
+	// validate argument
+	if env.pachAddress == "" {
+		return errors.New("cannot initialize pach client with empty pach address")
+	}
+	// Initialize pach client
+	return backoff.Retry(func() error {
+		var err error
+		env.pachClient, err = client.NewFromAddress(env.pachAddress)
+		if err != nil {
+			return fmt.Errorf("failed to initialize pach client: %v", err)
+		}
+		return nil
+	}, backoff.RetryEvery(time.Second).For(time.Minute))
+}
+
+func (env *ServiceEnv) initEtcdClient() error {
+	// validate argument
+	if env.etcdAddress == "" {
+		return errors.New("cannot initialize pach client with empty pach address")
+	}
+	// Initialize etcd
+	return backoff.Retry(func() error {
+		var err error
+		env.etcdClient, err = etcd.New(etcd.Config{
+			Endpoints: []string{env.etcdAddress},
+			// Use a long timeout with Etcd so that Pachyderm doesn't crash loop
+			// while waiting for etcd to come up (makes startup net faster)
+			DialOptions: append(client.DefaultDialOptions(), grpc.WithTimeout(3*time.Minute)),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to initialize etcd client: %v", err)
+		}
+		return nil
+	}, backoff.RetryEvery(time.Second).For(time.Minute))
+}
+
+func (env *ServiceEnv) initKubeClient() error {
+	return backoff.Retry(func() error {
+		// Get secure in-cluster config
+		var kubeAddr string
+		var ok bool
+		cfg, err := rest.InClusterConfig()
+		if err == nil {
+			goto connect
```

Comment:
Any particular reason we need goto here? It looks like you could just as easily put the error handling in an `if err != nil` block so to get the same semantic effect a little more cleanly.

Reply:
Yeah, that seems like a good idea. This is actually something I pulled straight from Matt's work, so I do not know the reason why a goto was used.